File: python/chapter_2/07_intersection.py
# Проверьте, пересекаются ли два заданных (одно-)связных списка.
# Верните узел пересечения. Учтите, что пересечение определяется ссылкой, а не значением.
# Иначе говоря, если k-й узел первого связного списка точно совпадает
# (по ссылке) с j-м узлом второго связного списка, то списки считаются пересекающимися
import unittest
from LinkedList import LinkedList


def intersection(l1, l2):
    if l1.last_node is not l2.last_node:
        return False

    diff = len(l1) - len(l2)
    if diff >= 0:
        long, short = l1, l2
    else:
        long, short = l2, l1

    node_short, node_long = short.start_node, long.start_node
    for i in range(abs(diff)):
        node_long = node_long.ref

    while node_short is not node_long:
        node_short, node_long = node_short.ref, node_long.ref

    return node_short


class Test(unittest.TestCase):
    def test_intersection_true(self):
        l1 = l2 = LinkedList()
        l1.insert_multiple(1, 2, 3)
        result = intersection(l1, l2)
        self.assertTrue(result)

    def test_intersection_false(self):
        l1 = LinkedList()
        l1.insert_multiple(1, 2, 3)
        l2 = LinkedList()
        l2.insert_multiple(4, 5, 6, 7)
        self.assertFalse(intersection(l1, l2))


if __name__ == '__main__':
    unittest.main()

File: go/chapter_2/07_intersection.go
// Проверьте, пересекаются ли два заданных (одно-)связных списка.
// Верните узел пересечения. Учтите, что пересечение определяется ссылкой, а не значением.
// Иначе говоря, если k-й узел первого связного списка точно совпадает
// (по ссылке) с j-м узлом второго связного списка, то списки считаются пересекающимися
package chapter_2

func Intersection(l1, l2 *LinkedList) (bool, *node) {
	if l1.tail != l2.tail {
		return false, nil
	}

	diff := l1.Len() - l2.Len()
	n1, n2 := l1.head, l2.head
	if diff > 0 {
		for i := 0; i < diff; i++ {
			n1 = n2.next
		}
	}
	if diff < 0 {
		diff = -diff
		for i := 0; i < diff; i++ {
			n2 = n2.next
		}
	}

	for n1 != nil {
		if n1 == n2 {
			return true, n1
		}
		n1 = n1.next
		n2 = n2.next
	}
	return false, nil
}
